product_sequence/models/product_product.py

```diff
@@ -47,20 +47,20 @@ def write(self, vals):
         Note this is up to the user, if the product category is changed,
         she/he will need to write '/' on the internal reference to force the
         re-assignment."""
-        if vals.get('default_code', '') in [False, '/']:
-            for product in self:
-                sequence = product.categ_id.sequence_id
+        for product in self:
+            if vals.get('default_code', '') == '/':
+                category_id = vals.get('categ_id', product.categ_id.id)
+                category = self.env['product.category'].browse(category_id)
+                sequence = category.sequence_id
                 if not sequence:
                     sequence = self.env.ref(
                         'product_sequence.seq_product_auto')
                 ref = sequence.next_by_id()
```

Comment:
could be added a logger info that ref was changed. Like "Internal Reference XX of product  YY was changed to ZZ"